### ISSUE TITLE
triplet loss

### ISSUE THREAD

**Issue by Niculuse** [huanghoujing/AlignedReID-Re-Production-Pytorch#16]
作者你好,！
       我发现在open-reid中公布的market1501上triplet loss的rank1有85%左右，我运行源码用一块GPU，batch_size为128，初始学习率没调整，150轮之后rank1只有79%多一点，然后我又把初始学习率减半了，batch_size还是128，结果还是79%多，过不了80%，和公布结果差不少，用softmax loss结果都有80.5%的rank1准确率，比公布的结果只低一丁点。但是到triplet loss这反而比softmax loss还低。我看您研究过open-reid中的triplet loss，发表过很多issues，所以想请教一下您。另外我用的Python2.7.3，pytorch0.2，open-reid0.2.0。不光在market1501上，在cuhk03上也是，比公布的结果低8个点，但是比。
       我结合分类损失和triplet loss在market1501上也出现了效果下降的现象，不知道作者您解决了吗？
感谢！

**Comment by huanghoujing** [comment]
open-reid，batch size 256可以跑到rank-1 84.5%，[参考这里](https://github.com/Cysu/open-reid/issues/15). 印象中应该是在open-reid划分的Market1501上跑的实验，而不是在[标准划分](https://github.com/Cysu/open-reid/issues/16)上跑的。

虽然我用Python 2跑的，你可以换成Python 3试一下，如果还不行那有可能跑出来就是那样吧。

也可以参考一下我这个工程里的一些设定，和open-reid稍有不同。进一步[我做了一个小trick](https://github.com/huanghoujing/person-reid-triplet-loss-baseline)，可以跑到89%。

分类loss和triplet loss结合我还是没解决，你什么时候解决了跟我说一下，3q : )

**Comment by Niculuse** [comment]
我试过2块显卡跑256batch size 还是低了几点，跑不到那个水平。
谢谢你了！

**Comment by Phoebe-star** [comment]
hi 
how about your triplet loss?

above is my loss result , is it right ? the value is small?
Train loss is triplet loss
Step: 79, Learning rate: 0.009942, Train loss: 0.157365
('pos_loss :   ', 0.80680853)
('neg_loss :   ', 1.0257512)



Step: 5409, Learning rate: 0.007231, Train loss: 0.057204
('pos_loss :   ', 1.3745451)
('neg_loss :   ', 1.9384971)

**Comment by liliangqi** [comment]
@huanghoujing 您这里的trick具体是指什么呢

**Comment by huanghoujing** [comment]
@liliangqi 这里的trick就是把残差网络conv5的stride从2改为1，这样改过之后average pooling之前特征的分辨率就加倍了. 这个增加分辨率的做法是从文章*Beyond Part Models: Person Retrieval with Refined Part Pooling*看到的。

**Comment by huanghoujing** [comment]
@Phoebe-star I think your loss value `0.057204` is not small enough yet. You can also calculate how many triplets satisfy the margin inequality `distance(anchor, positive) + margin < distance(anchor, negative)`, which I think is a better indicator.

**Comment by Niculuse** [comment]
@huanghoujing 大神有没有看过Harmonious Attention Network for Person Re-identification 这篇文章，如果有的话，我想请教些问题。

**Comment by Phoebe-star** [comment]
Step: 5409, Learning rate: 0.007231,  Local Train loss: 0.29
('local pos_loss : ', 12.745451)
('local neg_loss : ', 11.384971)

I find the  local  feature  doesn't   effect . 
the local feature use the 
![image](https://user-images.githubusercontent.com/29423761/37640115-180bf252-2c4f-11e8-820a-30156f048782.png)

S (7 ,7) is the final distance , but its value is big , like  12.46782  or 11.14258
I think  your local feature ,it is doing 7x7 times, because you use  the  " for  loop" (python) ,
but I  watch the image 
![image](https://user-images.githubusercontent.com/29423761/37640380-38dff0d6-2c50-11e8-982f-41aedd366ca1.png)

it seem like do 7+6+5+4+3+2+1=28 times

answer can English or 中文 ,thanks

**Comment by huanghoujing** [comment]
@Phoebe-star The paper explains how to calculate the shortest distance well, and I just follow the paper. So any other explanation I would like to give here is just a duplicate. Maybe I wouldn't be able to get things right for you. Sorry for that.

**Comment by Niculuse** [comment]
@huanghoujing 好的，感谢！

**Comment by huanghoujing** [comment]
If this value `12.745451` is loss, then it's large.

**Comment by Phoebe-star** [comment]
how about your local loss?

**Comment by huanghoujing** [comment]
Similar to global loss, local loss can also decreases to near 0, e.g. 0.0003.

**Comment by Phoebe-star** [comment]
where is the local  loss compute method ?  I find you design two method . one is numpy ,another is torch

**Comment by Phoebe-star** [comment]
https://github.com/huanghoujing/AlignedReID-Re-Production-Pytorch/issues/2
hi ,you say  " distance of two unit-length vectors falls in range [0, 2],  "
why not in the [0,1]? because normalization


and why  after Equation (1), the distance is in range [0, 0.76]? 


thanks you , god

**Comment by JackeyWang777** [comment]
您好，请问一下分类loss和triplet loss结合会遇到什么问题呢？

**Comment by huanghoujing** [comment]
两个loss结合理论上会更好。但是我在Triplet Loss的performance已经很高的设定下再加分类loss，结果并没有提升，我弄不好这个。

**Comment by Gavin666Github** [comment]
Combining softmax loss with metric learning loss to speed up the convergence is also a popular method.
